src/main/java/rx/internal/operators/OperatorBufferPredicateBoundary.java

```diff
@@ -0,0 +1,487 @@
+import rx.internal.util.unsafe.*;
+
+/**
+ * Buffers values into a continuous, non-overlapping Lists where the boundary is determined
+ * by a predicate returning true.
+ *
+ * @param <T> the source and List element type
+ */
+public final class OperatorBufferPredicateBoundary<T> implements Operator<List<T>, T> {
+
+    final Func1<? super T, Boolean> predicate;
+    
+    final int prefetch;
+    
+    final int capacityHint;
+    
+    final boolean after;
+    
+    public OperatorBufferPredicateBoundary(Func1<? super T, Boolean> predicate, int prefetch, int capacityHint, boolean after) {
+        if (predicate == null) {
+            throw new NullPointerException("predicate");
+        }
+        if (prefetch <= 0) {
+            throw new IllegalArgumentException("prefetch > 0 required but it was " + prefetch);
+        }
+        if (capacityHint <= 0) {
+            throw new IllegalArgumentException("capacityHint > 0 required but it was " + capacityHint);
+        }
+        this.predicate = predicate;
+        this.prefetch = prefetch;
+        this.capacityHint = capacityHint;
+        this.after = after;
+    }
+
+    @Override
+    public Subscriber<? super T> call(Subscriber<? super List<T>> child) {
+        if (after) {
+            final BoundedAfterSubscriber<T> parent = new BoundedAfterSubscriber<T>(child, capacityHint, predicate, prefetch);
+            child.add(parent);
+            child.setProducer(new Producer() {
+                @Override
+                public void request(long n) {
+                    parent.requestMore(n);
+                }
+            });
+            return parent;
+        }
+        final BoundedBeforeSubscriber<T> parent = new BoundedBeforeSubscriber<T>(child, capacityHint, predicate, prefetch);
+        child.add(parent);
+        child.setProducer(new Producer() {
+            @Override
+            public void request(long n) {
+                parent.requestMore(n);
+            }
+        });
+        return parent;
+    }
+    
+    static final class BoundedAfterSubscriber<T> extends Subscriber<T> {
+        final Subscriber<? super List<T>> actual;
+        
+        final int capacityHint;
+        
+        final Func1<? super T, Boolean> predicate;
+        
+        final Queue<Object> queue;
+        
+        final AtomicLong requested;
+
+        final AtomicInteger wip;
+        
+        final NotificationLite<T> nl;
+        
+        final int limit;
+
+        List<T> buffer;
+        
+        long upstreamConsumed;
+        
+        volatile boolean done;
+        Throwable error;
+
+        public BoundedAfterSubscriber(Subscriber<? super List<T>> actual, int capacityHint,
+                Func1<? super T, Boolean> predicate, int prefetch) {
+            this.actual = actual;
+            this.capacityHint = capacityHint;
+            this.predicate = predicate;
+            Queue<Object> q;
+            if (prefetch >= 10 * 1000 * 1000) {
```

Comment:
It's not time, it's prefetch amount. Also, your variant will require runtime calculation, David's may be calculated at compile time (nano win!).